### DISCUSSION TITLE
[RFC5746] Secure server-side renegotiation is still NOT supported (?)

### DISCUSSION THREAD

**Discussion by nser77** [openssl/openssl#21747] [Q&A] [answered]
Hi all, I hope someone can clarify me this behavior.

I tried to built my nginx/openresty web server against OpenSSL 3.1.2 to fix the RFC5746 vulnerability; after upgrading from 1.1.1 branch to 3.0.0 branch, **secure server-side renegotiation** is still NOT supported.

As per my understanding, this could happens with branch 1.1.1, but in 3.0.0 the **secure renegotiation** option is enabled by default.

Where am I wrong?
Do I need to check the code from the web-server upstream repository?

Many thanks

----

OpenSSL details: 

```
OpenSSL 3.1.2 1 Aug 2023 (Library: OpenSSL 3.1.2 1 Aug 2023)
built on: Tue Aug 15 22:23:15 2023 UTC
platform: linux-x86_64
options:  bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O0 -g -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL
OPENSSLDIR: "/usr/local/ssl"
ENGINESDIR: "/usr/local/ssl/lib64/engines-3"
MODULESDIR: "/usr/local/ssl/lib64/ossl-modules"
Seeding source: os-specific
CPUINFO: OPENSSL_ia32cap=0x80202001478bfffd:0x0

```

nginx/openresty details:

```
nginx version: openresty/1.21.4.2
built by gcc 8.3.0 (Debian 8.3.0-6)
built with OpenSSL 3.1.2 1 Aug 2023
TLS SNI support enabled
configure arguments: --prefix=/etc/nginx/nginx --with-cc-opt='-O2 -I/usr/local/ssl/include' --add-module=../ngx_devel_kit-0.3.2 --add-module=../echo-nginx-module-0.63 --add-module=../xss-nginx-module-0.06 --add-module=../ngx_coolkit-0.2 --add-module=../set-misc-nginx-module-0.33 --add-module=../form-input-nginx-module-0.12 --add-module=../encrypted-session-nginx-module-0.09 --add-module=../srcache-nginx-module-0.33 --add-module=../ngx_lua-0.10.25 --add-module=../ngx_lua_upstream-0.07 --add-module=../headers-more-nginx-module-0.34 --add-module=../array-var-nginx-module-0.06 --add-module=../memc-nginx-module-0.19 --add-module=../redis2-nginx-module-0.15 --add-module=../redis-nginx-module-0.3.9 --add-module=../rds-json-nginx-module-0.16 --add-module=../rds-csv-nginx-module-0.09 --add-module=../ngx_stream_lua-0.0.13 --with-ld-opt='-Wl,-rpath,/etc/nginx/luajit/lib -L/usr/local/ssl/lib64' --sbin-path=/usr/sbin/nginx --modules-path=/usr/lib/nginx/modules --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp --http-scgi-temp-path=/var/cache/nginx/scgi_temp --user=nginx --group=nginx --with-compat --with-threads --with-http_addition_module --with-http_auth_request_module --with-http_dav_module --with-http_flv_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_mp4_module --with-http_random_index_module --with-http_realip_module --with-http_secure_link_module --with-http_slice_module --with-http_ssl_module --with-http_stub_status_module --with-http_sub_module --with-http_v2_module --with-mail --with-mail_ssl_module --with-stream --with-stream_realip_module --with-stream_ssl_module --with-stream_ssl_preread_module --with-stream
```

**secure server-side renegotiation** is still NOT supported:

```
openssl s_client -port 443 mydomain.com

[...]

---
SSL handshake has read 2819 bytes and written 402 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 384 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)

[...]

```

**Comment by t8m** [comment, marked as answer]
The secure renegotiation is enabled by default already on 1.1.1 branch.

The message is misleading because TLSv1.3 was negotiated and that version of TLS does not support any renegotiation at all so neither secure nor insecure renegotiation is supported.

**Reply by nser77** [reply]
Many thanks.

I just checked the RFC8446 and now all make sense.

**Reply by nser77** [reply]
I checked the source code of **OpenSSL** and seems that secure renegotiation warning should be already suppressed when a TLS1.3 handshake occurs; from ```openssl/apps/lib/s_cb.c```:

```
void ssl_print_secure_renegotiation_notes(BIO *bio, SSL *s)
{
    if (SSL_VERSION_ALLOWS_RENEGOTIATION(s)) {
        BIO_printf(bio, "Secure Renegotiation IS%s supported\n",
                   SSL_get_secure_renegotiation_support(s) ? "" : " NOT");
    } else {
        BIO_printf(bio, "This TLS version forbids renegotiation.\n");
    }
}
```

I'm trying to debug it and find out a solution.

**Reply by davidben** [reply]
> The message is misleading because TLSv1.3 was negotiated and that version of TLS does not support any renegotiation at all so neither secure nor insecure renegotiation is supported.

I think this is the wrong value for OpenSSL to return. BoringSSL intentionally makes `SSL_get_extms_support` and `SSL_get_secure_renegotiation_support` return true in TLS 1.3. This was based on examining the actual use cases for those functions. It's not "can I do a secure renegotiation?" but "does the connection have the corresponding security fixes applied?". In TLS 1.3, the answer to the latter question is yes.

An application that wishes to use, say, a channel binding while being secure against the [3SHAKE](https://www.mitls.org/pages/attacks/3SHAKE) attack would need to check `SSL_get_extms_support` and reject the connection if false. In BoringSSL's semantics, it is sufficient to call the accessor. [tokbind](https://www.rfc-editor.org/rfc/rfc8471.html#section-4.2) is an example of a protocol that expects this. In OpenSSL's semantics, the application needs to know quirks of how the bugfix applies to TLS versions and correlate with a version check. This gets additionally complicated if the application supports both DTLS and TLS because the version numbers are different.

Indeed, these are the semantics even in the protocol. Even in TLS 1.2, `SSL_get_secure_renegotiation_support` returning true does *not* imply that secure renegotiation is supported. It only implies that all renegotiations on this connection are secure, i.e. no insecure renegotiations will happen on this connection. It is perfectly valid, and indeed expected, for a server that disallows all renegotiation to still negotiate the extension. See the discussion in https://github.com/openssl/openssl/issues/21285 and https://github.com/openssl/openssl/discussions/21437/. In that case, `SSL_get_secure_renegotiation_support` will return true, on both client and server, despite the server rejecting all renegotiations. Moreover, the client has no way of knowing the server will do this because it's not signaled in the handshake. OpenSSL couldn't report on whether secure renegotiation will work even if it wanted.

Under those corrected semantics, it becomes clear that `SSL_get_secure_renegotiation_support` should return true for TLS 1.3. No insecure renegotiations will occur in a TLS 1.3 connection. No secure renegotiations will happen either, but `SSL_get_secure_renegotiation_support` already could not say anything useful about it in TLS 1.2.

**Comment by nser77** [comment]
> The secure renegotiation is enabled by default already on 1.1.1 branch.

Sorry again.

Based on your input, I tried to connect to an RFC5746 vulnerable web-server with `OpenSSL 1.1.1n`;  in that case the TLS handshake was performed and I got the following result:

```
CONNECTED(00000003)

[...]

---
SSL handshake has read 5068 bytes and written 443 bytes
Verification: OK
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 4096 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated

---

```

When I trying to connect to the same vulnerable web server with `OpenSSL 3.1.2`, there is no TLS handshake and the connection is aborted:

```
CONNECTED(00000003)
00C14C2A837F0000:error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation disabled:ssl/statem/extensions.c:893:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 85 bytes and written 334 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
---
```

Why is the behavior of `OpenSSL 1.1.1n` different than `OpenSSL 3.1.2` against a vulnerable RFC5746 web-server?
Based on your answer, should `OpenSSL 1.1.1n` also abort the connection?


Many thanks

**Reply by mattcaswell** [reply]
What @t8m said was:
> The secure renegotiation is enabled by default already on 1.1.1 branch.

This is entirely correct. Secure renegotiation will always be *requested* by an OpenSSL client and will always be *required* by an OpenSSL server in 1.1.1. If a client that does not support secure reneg attempts to connect to a 1.1.1 server then it will fail.

However, if a 1.1.1 client attempts to connect to a server that does not support secure reneg then it will succeed, i.e. an 1.1.1 client always requests secure reneg but does not require it.

What changed with 3.0 is PR #15127. This changed the default case so that an OpenSSL client now requires a patched server for the handshake to succeed.

**Reply by nser77** [reply]
Many thanks for your answer.

Some server manufacturers have not yet patched their servers and from a client side prospective there are a lot of operative systems that ship **OpenSSL 1.1.1**; from my understanding, this results in possible attack scenarios.

RFC5746  [`Section 3.4 Client Behavior: Initial Handshake` ](https://www.rfc-editor.org/rfc/rfc5746#section-3.4) says that `some clients may want to terminate the handshake instead of continuing` but from a **SSL Library** prospective there is no reason to do that and the client MUST abort the handshake as per default behavior.

```
   o  When a ServerHello is received, the client MUST check if it
      includes the "renegotiation_info" extension:

      *  If the extension is not present, the server does not support
         secure renegotiation; set secure_renegotiation flag to FALSE.
         In this case, some clients may want to terminate the handshake
         instead of continuing; see Section 4.1 for discussion.
```

[`Section 4.1`](https://www.rfc-editor.org/rfc/rfc5746#section-4.1) confirms that if a client wants to prevent this attack, it needs to terminate the connection.

```
   If clients wish to ensure that such attacks are impossible, they need
   to terminate the connection immediately upon failure to receive the
   extension without completing the handshake.  Such clients MUST
   generate a fatal "handshake_failure" alert prior to terminating the
   connection.  However, it is expected that many TLS servers that do
   not support renegotiation (and thus are not vulnerable) will not
   support this extension either, so in general, clients that implement
   this behavior will encounter interoperability problems.  There is no
   set of client behaviors that will guarantee security and achieve
   maximum interoperability during the transition period.  Clients need
   to choose one or the other preference when dealing with potentially
   un-upgraded servers.
```

From an **SSL Library** prospective, I'd expect to enforce the RFC5746 and **require** secure renegotiation; I think **OpenSSL 1.1.1** MUST prevent this vulnerability: seems that most developers just delegate the handshake to the **SSL library** so there are many potentially vulnerable (client) applications.

Finally, in most cases, a server is also acting as a client, by calling an API endpoint for example, so from my understanding this could result in a possible attack scenario; Am I wrong with this concern?

**Reply by mattcaswell** [reply]
As noted in RFC5746 section 4.1 there are are a number of different considerations as to what should happen in the event of the server failing to support secure renegotiation. As far as the client is concerned it is optional whether to proceed or not. There is a classic interoperability/security trade off to be made here. When RFC5746 first came out it was impractical to require clients to abort the handshake by default if the server did not support secure renegotiation - because most servers did not do so. OpenSSL's implementation of this behaviour dates from this time. It was judged at that time sufficient for clients to always offer it, but to continue with the handshake anyway if the server did not support it. Of course over time the trade off changes as the prevalence of secure renegotiation changes. Which is why the default was changed in 3.0. Backporting such a change to a stable version is breaking so we try to reserve such changes to new releases only. Given that 1.1.1 is EOL in less than 4 weeks time there is zero chance that it will be changed now.
